cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/RepositoryManager.java

```diff
@@ -0,0 +1,250 @@
+    this.validationRepoDirectory = repoBasePath.resolve(VALIDATION_REPOSITORY_DIRECTORY_NAME).toFile();
+    try {
+      this.authStrategy = getAuthStrategy();
+    } catch (AuthStrategyNotFoundException e) {
+      //  This exception isn't expected as the API handlers only store auth config after validations,
+      throw new RuntimeException(e);
+    }
+  }
+
+  private <C extends TransportCommand> C getAuthenticatedCommandWithTimeout(C command) throws AuthenticationException {
+    command.setCredentialsProvider(authStrategy.getCredentialProvider(secureStore, repositoryConfig, namespaceId));
+    command.setTimeout(gitCommandTimeoutSeconds);
+    return command;
+  }
+
+  /**
+   * Initializes the Git repository by cloning remote.
+   *
+   * @throws GitAPIException         when a Git operation fails.
+   * @throws AuthenticationException when correct authentication credentials are not available.
+   * @throws IOException             when file or network I/O fails.
+   */
+  public void initialize() throws GitAPIException, AuthenticationException, IOException {
+    if (git != null) {
+      return;
+    }
+    // Clean up the directory if it already exists.
+    deleteDirectoryIfExists(localRepoDirectory);
+    CloneCommand command =
+      Git.cloneRepository().setURI(repositoryConfig.getLink()).setDirectory(localRepoDirectory);
+
+    String branch = getBranchRefName(repositoryConfig.getDefaultBranch());
+    if (branch != null) {
+      command.setBranchesToClone(Collections.singleton((branch))).setBranch(branch);
+    }
+    command = getAuthenticatedCommandWithTimeout(command);
+    command.call();
+  }
+
+  // Returns the ref name for the branch or Null if the branch name is null.
+  @Nullable
+  private String getBranchRefName(@Nullable String branch) {
+    if (Strings.isNullOrEmpty(branch)) {
+      return null;
+    }
+    return "refs/heads/" + branch;
+  }
+
+  /**
+   * Returns an {@link AuthStrategy} for the given Git repository provider and auth type.
+   *
+   * @return an instance of {@link  AuthStrategy}.
+   * @throws AuthStrategyNotFoundException when the corresponding {@link AuthStrategy} is not found.
+   */
+  private AuthStrategy getAuthStrategy() throws AuthStrategyNotFoundException {
+    RepositoryConfig config = this.repositoryConfig;
+    if (config.getProvider() == Provider.GITHUB) {
+      if (config.getAuth().getType() == AuthType.PAT) {
+        return new GitPATAuthStrategy();
+      }
+    }
+    throw new AuthStrategyNotFoundException(String.format("No strategy found for provider %s and type %s.",
+                                                          config.getProvider(),
+                                                          config.getAuth().getType()));
+  }
+
+  @Nullable
+  private RepositoryValidationFailure validateDefaultBranch(Collection<Ref> refs) {
+    // If refs are null, i.e. couldn't be fetched or
+    if (refs == null) {
```

Comment:
Will this happen? If so, annotate the parameter with `@Nullable`. Otherwise no need to have this check.